src/main/scala/chisel3/interface/Interface.scala

```diff
@@ -0,0 +1,121 @@
+  private[interface] def portMap: Seq[(Mod, Intf#Ports) => (Data, Data)]
+
+}
+
+/** Functionality which is common */
+sealed trait InterfaceCommon {
+
+  private[interface] type Ports <: Record
+
+  /** Returns the Record that is the port-level interface. */
+  private[interface] def ports(): Ports
+
+}
+
+/** A generator of different Interfaces. Currently, this is just an Interface
+  * that is not a Singleton.
+  */
+trait InterfaceGenerator extends InterfaceCommon
+
+/** An interface between hardware units. Any module that implements this
+  * interface may be separately compiled from any module that instantiates this
+  * interface.
+  */
+trait Interface extends InterfaceCommon { self: Singleton =>
+
+  /** This types represents the type of a valid conformance to this Interface.
+    */
+  private type Conformance[Mod <: RawModule] = ConformsTo[this.type, Mod]
+
+  /** The name of this interface. This will be used as the name of any module
+    * that implements this interface.
+    * I.e., this is the name of the `BlackBox` and `Module` that are provided
+    * below. The implementation of this method is just coming up with a good
+    * name derived from the class name. (The name of the Interface in FIRRTL
+    * will be the name of the Scala class that extends the Interface.)
+    */
+  private[interface] def interfaceName: String = {
+    val className = getClass().getName()
+    className
+      .drop(className.lastIndexOf('.') + 1)
+      .split('$')
+      .toSeq
+      .filter {
+        case str if str.forall(_.isDigit) => false
+        case str                          => true
+      }
+      .last
+  }
+
+  sealed trait Entity { this: BaseModule =>
+    val io: Ports
+  }
+
+  object Wrapper {
+
+    /** The black box that has the same ports as this interface. This is what is
+      * instantiated by any user of this interface, i.e., a test harness.
+      */
+    final class BlackBox extends chisel3.BlackBox with Entity {
+      final val io = IO(ports())
+
+      override final def desiredName = interfaceName
+    }
+
+    /** The module that wraps any module which conforms to this Interface.
+      */
+    final class Module[B <: RawModule](
+    )(
+      implicit conformance: Conformance[B])
+        extends RawModule
+        with Entity {
+      final val io = FlatIO(ports())
+
+      // Use a dummy clock and reset connection when constructing the module.
+      // This is fine as we rely on DataView to catch missing connections to
+      // clock and reset.  The dummy clock and reset will never be used.
+      private val internal = withClockAndReset(
+        WireInit(Clock(), DontCare),
+        WireInit(Reset(), DontCare)
+      ) {
+        chisel3.Module(conformance.genModule())
+      }
+
+      private implicit val pm = PartialDataView[B, Ports](
+        _ => ports(),
+        conformance.portMap: _*
+      )
+      io :<>= internal.viewAs[Ports]
```

Comment:
curious will this error if any ports of the Module were missed in this?

Reply:
The actual `io :<>= internal.viewAs[Ports]` connection should always be correct as this is connection `Ports :<>= Ports`. If the view is non-total, then an exception is thrown. I added some checking and a test in https://github.com/chipsalliance/chisel/pull/3199/commits/34cf05bd7c3a7707c92cb528eda166d941d88eff that makes it more obvious that the user has an invalid conformance.